src/Microsoft.AspNetCore.SignalR.Core/HubConnectionContext.cs

```diff
@@ -74,44 +74,110 @@ public HubConnectionContext(ConnectionContext connectionContext, TimeSpan keepAl
 
         public int? LocalPort => Features.Get<IHttpConnectionFeature>()?.LocalPort;
 
-        public virtual async Task WriteAsync(HubMessage message)
+        public virtual ValueTask WriteAsync(HubMessage message)
         {
-            await _writeLock.WaitAsync();
+            // We were unable to get the lock so take the slow async path of waiting for the semaphore
+            if (!_writeLock.Wait(0))
+            {
+                return new ValueTask(WriteSlowAsync(message));
+            }
+
+            // This method should never throw synchronously
+            var task = WriteCore(message);
+
+            // The write didn't complete synchronously so await completion
+            if (!task.IsCompletedSuccessfully)
+            {
+                return new ValueTask(CompleteWriteAsync(task));
+            }
+
+            // Otherwise, release the lock
+            _writeLock.Release();
```

Comment:
Where is this writeLock acquired?

Reply:
Ah, it is the `Wait` on line 80. Could you add a little bit of documentation, e.g.

`// Otherwise, release the lock acquired when entering WriteAsync`